### DISCUSSION TITLE
Cache update design for batch updates

### DISCUSSION THREAD

**Discussion by trompx** [urql-graphql/urql#1417] [Q&A] [answered]
Hello,

I have an app where when doing a specific action, I have to batch update multiple items (few dozen to hundreds).
I just send the id of the entry to update and the modified property value, and return just that. I don't want after the update to query again all the updated items to return their whole data so that the cache can be automatically updated, as the items are complex and require multiple joins so I want to leverage the data already fetched and availabled in the app and manually update the cache. 

In Apollo, I could directly at the mutation execution time specify an update function, thus having access to the data being processed. As I only send the id and the proprety value to be updated in the mutation, in the cacheExchange, I don't have access to the whole items data (as said above I don't return that from the batch update query and don't want to) to update the cache with writeFragment, so I need to create a cache/store to access that in the cacheExchange.

Could it be possible to implement something similar to apollo update so I don't have to create a cache/store?

**Comment by kitten** [comment, marked as answer]
All of the information on Graphcache's Cache Updates can be found on this page: https://formidable.com/open-source/urql/docs/graphcache/cache-updates/

It has some concrete guides on multiple cases like the one you're describing.

If you post more information I could even maybe help you write out some examples for your exact use-case.

As you haven't posted much, I obviously can't check whether there's some kind of misunderstand that I need to clear up. You did mention `cache.writeFragment` which is indeed one way to update entities but you say you don't have all information? Are you aware that you don't have to write the full fragment but only the information that you need to update?

We also don't add updates to your UI code because we see that as an anti-pattern. It encourages multiple patterns that couple data / state management code to your UI and typically leads to bad maintainability. In this case, if we had it, you'd probably choose to use the data from a random query to supplement data that the cache has anyway and some may use it to diverge from having the cache hold data that stays true to their server data.

Anyway, I can probably help with some examples that are closer to your code ✌️

**Comment by trompx** [comment]
Thanks a lot for proposing your help @kitten.
I already read the cache updates and optimistic updates docs so maybe I am missing something.

I first fetch some data `fetchedLists` and modify it on the UI -> `fetchedListsModified` (with dozen to hundreds of modifications).
Then I batch update all modifications:
```ts
batchUpdateMutation({
  data: {
    items: fetchedListsModified.map(
      ({ id, listId, propertyToModifyValue }) => ({
        id,
        listId,
        property: propertyToModifyValue,
      })
    ),
  },
});
```
If I execute that mutation without doing any update in cacheResolver, the UI doesn't update (which is normal as I don't return full item object type).

My batchUpdate query return data like (as I said I don't query the full object type after the update as it's a heavy query and the data has already been fetched):
```ts
[
  {id: 1, listId: 1, property: "..."},
  {id: 2, listId: 2, property: "..."}
]
```

In cacheExchange updates, I do something like:

```ts
  cache.writeFragment(ListFragmentDoc, {
    id: listId,
    // I need the full items here, not just the data I return from the update mutation
    // -> that is why I pass that data (fetchedListXModified is a filtered part of fetchedListModified by listId)
    // via a cache/store from the component which fetched the data, as the cached data
    // doesn't have the modification
    items: fetchedListXModified,
  });
```
And ok, my question was more about the design and why Apollo allowed to update directly from the UI, so thanks for the clarification on that part.

**Reply by kitten** [reply]
Ok, so in essence you know how to update the items, but you don't know how to retrieve additional data, right? What might work is that you may want to read the fragment before writing it fragment if I'm not mistaken? https://formidable.com/open-source/urql/docs/graphcache/local-resolvers/#reading-a-fragment

**Reply by trompx** [reply]
Yes, basically I was wondering if there was a better way to get the value being modified in my component in the cacheExchange, without passing it by a custom cache/store as I don't return it from the server (because heavy query). And if I read the fragment in the cacheExchange, it gets the data before modification.

Edit: I also do an optimistic update direcly on the Items of the list so maybe I could get the data from there, but not sure if this is possible and cleaner.

**Reply by kitten** [reply]
Just to summarise this; you can't get the modifications though either way, even if you had an updater inside your UI logic / components / wherever you start the mutation, since those are performed by the API. So I'm not quite sure what you're referencing here.

I'd assume your mutation must receive enough information in your variables to actually perform this change, otherwise the API obviously wouldn't know what to do. So what stops you from combining the information that the cache gives you (i.e. before the modifications) with the information you have from the mutation and the mutation's variables?

**Reply by trompx** [reply]
Oh ok, I think I get what you're saying. You mean I would do a readFragment on the list cache, then do a writeFragment, loop the list cached items and replace with what I return from the server.

In my case, an item in one list can also be moved to another list during the update, which require now to manage add and remove from cache. If one of the item returned by the server is not in the current list, I need to get the item data from cache, update it and append it to list cache. I see now it is possible with just what is already in cache + what is returned from the server, but I have to **reproduce the add/remove from list logic** which have **already been done** in the UI which is not really convenient...

**Reply by kitten** [reply]
Ok, so this sounds like no matter where you put the update, you can't predict exactly what will happen. At least not without a huge amount of logic. That being said, if your logic is sent to the API then you're already doing some work to figure out what changes are happening. Of course I don't know what your mutation looks like but it sounds a little like a kanban board.

If the variables for moving the items around is too abstract and you control the API maybe there's an easier way to structure those? However, maybe having a `cache.updateQuery` call could also be easier to handle than individual fragments?

Writing arbitrary data from the UI to the cache can seem like a good idea, however it kind of seems like there's already a strict separation of what the UI and the server are doing.

What I've seen done before is that this state could be moved entirely to Graphcache. In such a case you wouldn't handle the state and batches separately. In essence, you would use optimistic updates to make the changes instant and not deal with batches until you send the mutations. This would require a custom exchange however to batch the changes.

So primarily, what I'm saying is, all solutions, especially if you want this to be optimistic, would require you to implement this update. You can "cheat" of course and work around this by keeping a `Map` or other store of the UI state around but essentially the difference between the two approaches above is that you either make granular changes or you code the same granular changes in a loop.

What kind of updates are you dealing with? If the only updates are moving items around between lists then in theory the update could be quite simple.

**Reply by trompx** [reply]
Yes, it's similar to a kanban board, so I need to handle position among other things but it's still manageable. I'm just trying to not repeat the logic of adding/removing item from list.

If I pass the data from the UI, I just have to make 2 cache.writeFragment. Updating the query fetching all the list wouldn't be simpler as I would need to update the 2 lists that have changed (so basically the same logic as with writeFragments).

If I have some problem with my current setup I may give a shot to the custom exchange.

Thanks again for your time and willingness to help, much appreciated.
